planner/core/expression_rewriter.go

```diff
@@ -670,9 +638,40 @@ func (er *expressionRewriter) handleInSubquery(v *ast.PatternInExpr) (ast.Node,
 		er.err = errors.Trace(err)
 		return v, true
 	}
-	er.p, er.err = er.b.buildSemiApply(er.p, np, expression.SplitCNFItems(checkCondition), asScalar, v.Not)
-	if er.err != nil {
-		return v, true
+	// If it's not the form of `not in (SUBQUERY)`, has no correlated column and don't need to append a scalar value. We can rewrite it to inner join.
+	if er.ctx.GetSessionVars().AllowInSubqToJoinAndAgg && !v.Not && !asScalar && len(np.extractCorrelatedCols()) == 0 {
+		// We need to try to eliminate the agg and the projection produced by this operation.
+		er.b.optFlag |= flagEliminateAgg
+		er.b.optFlag |= flagEliminateProjection2
+		// Build distinct for the inner query.
+		agg := er.b.buildDistinct(np, np.Schema().Len())
+		for _, col := range agg.schema.Columns {
+			col.IsAggOrSubq = true
+		}
+		eq, left, right, other := extractOnCondition(expression.SplitCNFItems(checkCondition), er.p, agg, false, false)
+		// Build inner join above the aggregation.
+		join := LogicalJoin{
+			JoinType:        InnerJoin,
+			EqualConditions: eq,
+			LeftConditions:  left,
+			RightConditions: right,
+			OtherConditions: other,
+		}.init(er.ctx)
+		join.SetChildren(er.p, agg)
+		join.SetSchema(expression.MergeSchema(er.p.Schema(), agg.schema))
```

Comment:
The schema of join should be outer children schema since it is originally a semi join?